### PR TITLE
Tokens metadata

### LINKED ISSUE #194: Fetch incoming & outgoing mempool txs
Cf. https://github.com/alephium/explorer-backend/pull/465

### PR DESCRIPTION
**Pull request by mvaivre**
Fixes #194

### CODE REVIEW

**Review by nop33 — COMMENTED** [review]
Lot's of good work!! 👏 

In my opinion, the explorer codebase could use some more love from our side, we just need to check with our priorities. Our page components have started growing a lot, and they include a lot of async logic in multiple `useEffects`, with many API calls whose results are lost once we navigate away from each page.

So far I think the requirements of the explorer were (and still are) more minimal than the ones from our other apps, but I feel that the introduction of tokens starts bloating the current architecture.

File: src/api/addressApi.ts [hunk 1]
```diff
@@ -16,27 +16,40 @@ You should have received a copy of the GNU Lesser General Public License
...
+  try {
+    const tokenIds = await clients.explorer.addresses.getAddressesAddressTokens(addressHash)
+
+    const tokens = await Promise.all(
+      tokenIds.map(async (id) => {
+        const tokenWithBalance = await clients.explorer.addresses.getAddressesAddressTokensTokenIdBalance(
+          addressHash,
+          id
+        )
+        const tokenMetadata = await getAssetMetadata({ assetId: id, networkType, nodeClient: clients.node })
```
[the remark this answers is not in the document]
Comment (mvaivre) [reply src/api/addressApi.ts:38]: I share your concerns. This is a good summary of what I started to see while coding - as said on our call, I know that this is not optimal. We've indeed probably reached the limits of our current architecture.

File: src/types/addresses.ts [hunk 2]
```diff
@@ -29,7 +30,7 @@ export type AddressBalanceResult = {
 
 export type AddressAssetsResult = {
   addressHash: AddressHash
-  assets: TokenBalances[]
+  assets: (AddressBalance & Partial<AssetInfo>)[]
```
[the remark this answers is not in the document]
Comment (mvaivre) [reply src/types/addresses.ts:33]: I looked at your types and adapted mines, but I indeed missed this one.

File: src/components/Links.tsx [hunk 3]
```diff
@@ -79,23 +80,28 @@ const AddressLinkBase = ({
...
+  useEffect(() => {
+    const getMetadata = async () => {
+      if (!clients?.node || !amounts) return
+
+      const metadata = await Promise.all(
+        amounts.map(async (a) => ({
+          ...(await getAssetMetadata({ assetId: a.id, networkType, nodeClient: clients?.node })),
+          ...a
+        }))
+      )
```
[the remark this answers is not in the document]
Comment (mvaivre) [reply src/components/Links.tsx:98]: I agree. I'll refactor this while working on the global state 🫠

File: src/pages/AddressInfoPage/index.tsx [hunk 4]
```diff
@@ -193,17 +194,18 @@ const AddressInfoPage = () => {
...
-    ...getAssetInfo({ assetId: a.id, networkType })
+    lockedBalance: BigInt(a.lockedBalance)
   })) ?? []) as Asset[]
 
   if (totalBalance && lockedBalance && parseInt(totalBalance) > 0) {
-    assets.push({ ...ALPH, balance: BigInt(totalBalance), lockedBalance: BigInt(lockedBalance) })
+    assets.push({ ...ALPH, balance: BigInt(totalBalance), lockedBalance: BigInt(lockedBalance), verified: true })
   }
 
+  assets = sortBy(assets, [(a) => !a.verified, 'name'])
```
[the remark this answers is not in the document]
Comment (mvaivre) [reply src/pages/AddressInfoPage/index.tsx]: Lol I freaking missed this one. Worked for too long.

**Comment by mvaivre** [comment]
I've decided to start working on the Redux switch, based on master. I'll reuse some of the code from this PR though.

**Pull request closed without merging**